staking-bank/src/lib.rs

```diff
@@ -0,0 +1,274 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+//! # Staking Bank
+use concordium_std::*;
+use core::fmt::Debug;
+
+#[cfg(feature = "production")]
+mod production_constants;
+#[cfg(feature = "production")]
+use production_constants::*;
+
+#[cfg(feature = "development")]
+mod development_constants;
+#[cfg(feature = "development")]
+use development_constants::*;
+
+#[cfg(feature = "sandbox")]
+mod sandbox_constants;
+#[cfg(feature = "sandbox")]
+use sandbox_constants::*;
+
+/// one = 1 * 10^18.
+#[cfg(any(feature = "production", feature = "development", feature = "sandbox"))]
+const ONE: u64 = 1000000000000000000u64;
+
+#[derive(Serial, Deserial)]
+pub struct State {}
+
+/// All smart contract errors.
+#[derive(Debug, PartialEq, Eq, Reject, Serial, SchemaType)]
+enum CustomContractError {
+    /// Failed to parse the parameter.
+    #[from(ParseError)]
+    ParseParams, // -1
+    /// Failed to log because the log is full.
+    LogFull, // -2
+    /// Failed to log because the log is malformed.
+    LogMalformed, // -3
+    /// Failed to invoke a contract.
+    InvokeContractError, // -4
+    /// Failed because the address is not a validator.
+    #[allow(dead_code)]
+    NotValidator, // -5
+}
+
+/// Mapping errors related to logging to CustomContractError.
+impl From<LogError> for CustomContractError {
+    fn from(le: LogError) -> Self {
+        match le {
+            LogError::Full => Self::LogFull,
+            LogError::Malformed => Self::LogMalformed,
+        }
+    }
+}
+
+/// Mapping errors related to contract invocations to CustomContractError.
+impl<T> From<CallContractError<T>> for CustomContractError {
+    fn from(_cce: CallContractError<T>) -> Self {
+        Self::InvokeContractError
+    }
+}
+
+/// Init function that creates a new smart contract.
+#[init(contract = "staking_bank")]
+fn init<S: HasStateApi>(
+    _ctx: &impl HasInitContext,
+    _state_builder: &mut StateBuilder<S>,
+) -> InitResult<State> {
+    Ok(State {})
+}
+
+/// Equivalent to solidity's getter function which is automatically created from the public storage variable `NUMBER_OF_VALIDATORS`.
+#[cfg(any(feature = "production", feature = "development", feature = "sandbox"))]
+#[receive(
+    contract = "staking_bank",
+    name = "NUMBER_OF_VALIDATORS",
+    return_value = "u8"
+)]
+fn number_of_validators<S: HasStateApi>(
+    _ctx: &impl HasReceiveContext,
+    _host: &impl HasHost<State, StateApiType = S>,
+) -> ReceiveResult<u8> {
+    Ok(NUMBER_OF_VALIDATORS)
+}
+
+/// Equivalent to solidity's getter function which is automatically created from the public storage variable `TOTAL_SUPPLY`.
+#[cfg(any(feature = "production", feature = "development", feature = "sandbox"))]
+#[receive(contract = "staking_bank", name = "TOTAL_SUPPLY", return_value = "u64")]
+fn total_supply_1<S: HasStateApi>(
+    _ctx: &impl HasReceiveContext,
+    _host: &impl HasHost<State, StateApiType = S>,
+) -> ReceiveResult<u64> {
```

Comment:
And u64 here is enough? What is the unit for this value, is it the same unit as ONE or in units of ONE?

Reply:
The highest value is currently 15*10^18 (production environment). I agree that if they intend to deploy many more validators in the future, they would need to use a bigger number. 

Nonetheless, the purpose of the `staking_bank` (as from some discussions) is currently only that 
"we just expect that balance of the validators to be not 0". In that sense, u64 is enough since all validators can be given the value "1" instead as well.

@DZariusz

Reply:
Note that "many more" is 3 more. That's the most you can do without overflowing with the current values.

Reply:
So should this be changed to a literal `1` instead of `10^18` per validator?

Reply:
Just a note:
The type for `ONE`, `balances`, and `total_supply` are all the same type `type StakingBalanceAmount = u64` and currently the `total_supply`/`balances` are units of ONE and hence the biggest values (where overflow can happen if more validators are added).

Reply:
I changed to use `1` instead of `10^18` as the `StakingAmountBalance=u8` value. The type allows for 255 validators before an overflow happens.

@DZariusz